.buildkite/autogenerate_pipeline.py

```diff
@@ -0,0 +1,240 @@
+import json
+import os
+import sys
+import pathlib
+import copy
+
+CONTAINER_VERSION = "v12"
+DOCKER_PLUGIN_VERSION = "v3.8.0"
+
+X86_AGENT_TAGS = os.getenv('X86_LINUX_AGENT_TAGS')
+AARCH64_AGENT_TAGS = os.getenv('AARCH64_LINUX_AGENT_TAGS')
+DOCKER_PLUGIN_CONFIG = os.getenv('DOCKER_PLUGIN_CONFIG')
+
+PARENT_DIR = pathlib.Path(__file__).parent.resolve()
+
+
+class BuildkiteStep:
+    """
+    This builds a Buildkite step according to a json configuration and the
+    environment variables `X86_LINUX_AGENT_TAGS`, `AARCH64_LINUX_AGENT_TAGS`
+    and `DOCKER_PLUGIN_CONFIG`. The output is a dictionary.
+    """
+
+    def __init__(self):
+        """
+        Initialize a Buildkite step with default values.
+        """
+
+        # Default values
+        self.label = None
+        self.command = None
+        self.retry = {'automatic': False}
+        self.agents = {'os': 'linux'}
+        self.plugins = [
+            {
+                f"docker#{DOCKER_PLUGIN_VERSION}": {
+                    'image': f"rustvmm/dev:{CONTAINER_VERSION}",
+                    'always-pull': True
+                }
+            }
+        ]
+
+    def _set_platform(self, platform):
+        """ Set platform if given in the json input. """
+
+        if platform:
+            if platform == 'aarch64':
+                platform = 'arm'
+            self.agents['platform'] = f"{platform}.metal"
+
+    def _set_conditional(self, conditional):
+        """ Set conditional if given in the json input. """
+
+        if conditional:
+            setattr(self, 'if', conditional)
+
+    def _add_docker_config(self, cfg):
+        """ Add configuration for docker if given in the json input. """
+
+        if cfg:
+            target = self.plugins[0][f"docker#{DOCKER_PLUGIN_VERSION}"]
+            for key, val in cfg.items():
+                target[key] = val
+
+    def _env_change_config(self, test_name, env_var, target, override=False):
+        """
+        Helper function to add to/override configuration of `target`
+        if `env_var` is set and this test appears in its list.
+        """
+
+        if env_var:
+            env_cfg = json.loads(env_var)
+
+            tests = env_cfg['tests']
+            assert tests,\
+                f"Environment variable {env_var} is missing the `tests` key."
+
+            cfg = env_cfg['cfg']
+            assert cfg,\
+                f"Environment variable {env_var} is missing the `cfg` key."
+
+            if test_name in tests:
+                if override:
+                    target.clear()
+                for key, val in cfg.items():
+                    target[key] = val
+
+    def _env_override_agent_tags(self, test_name):
+        """
+        Override the tags by which the linux agent is selected
+        using the `X86_LINUX_AGENT_TAGS` and `AARCH64_LINUX_AGENT_TAGS`
+        environment variables.
+        """
+
+        env_var = None
+        platform = self.agents.get('platform')
+
+        if platform:
```

Comment:
Can you add a comment about the logic here? It only becomes clear what you want to do when reading the build function. We should say that when no platform is specified, we don;t add the platform to the dictionary so that the test can run in any env.

Reply:
Done.